### PR TITLE
test: add profiler unit tests

### PR DESCRIPTION
**Pull request by dvilelaf**
## Proposed changes

- Add some profiler unit test
- Move the profiler wrap outside the loop to avoid redefinition

## Fixes

n/a

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes and CI passes too
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: aea/helpers/profiling.py [hunk 1]
```diff
@@ -114,22 +114,22 @@ def __init__(
...
+        def make_fn(obj: Any) -> Callable:
+            orig_new = obj.__new__
+            # pylint: disable=protected-access  # type: ignore
 
-                @wraps(orig_new)
-                def new(*args: Any, **kwargs: Any) -> Callable:
-                    self._counter[obj] += 1
-                    if orig_new is object.__new__:
-                        return orig_new(args[0])  # pragma: nocover
-                    return orig_new(*args, **kwargs)  # pragma: nocover
```
Comment (dvilelaf) [inline aea/helpers/profiling.py:129]: Moving this definition outside the loop.

File: tests/conftest.py [hunk 2]
```diff
@@ -590,6 +590,14 @@ def pytest_addoption(parser) -> None:
         help="check non closed threads i started during test",
     )
 
+    # Run profiler tests only
+    parser.addoption(
+        "--profiling",
+        action="store_true",
+        default=False,
+        help="run profiling tests only.",
+    )
```
Comment (dvilelaf) [inline tests/conftest.py:599]: If run with the `--profiling` flag, only profiler tests will run. If the flag is not present, all the other tests will be run.

File: tests/conftest.py [hunk 3]
```diff
@@ -1495,3 +1503,16 @@ def use_ipfs_daemon() -> Generator:
...
+    """Register the profiling marker to avoid warnings."""
+    config.addinivalue_line("markers", "profiling: filter profiler tests")
+
+
+def pytest_collection_modifyitems(config, items):
+    """Run profiler or non-profiler tests separately."""
+    if config.getoption("--profiling"):
+        items[:] = filter(lambda item: item.get_closest_marker("profiling"), items)
+    else:
+        items[:] = filter(lambda item: not item.get_closest_marker("profiling"), items)
```
Comment (dvilelaf) [inline tests/conftest.py:1518]: Here we select which set of tests we will run: profiler or the other ones.

File: tests/test_helpers/test_profiling.py [hunk 4]
```diff
@@ -18,13 +18,40 @@
...
+            )
+            continue
+
+        match = re.match(r".*\* (?P<field>.*) \(created\):  (?P<count>\d+).*", line)
+        if match:
+            result["created"][match.groupdict()["field"]] = int(
+                match.groupdict()["count"]
+            )
+
+    return result
```
Comment (dvilelaf) [inline tests/test_helpers/test_profiling.py:51]: Helper function that extracts instance counts for present and created objects so it is easier and cleaner to check numbers.

File: tests/conftest.py [hunk 5]
```diff
@@ -1495,3 +1503,16 @@ def use_ipfs_daemon() -> Generator:
 
     yield
     ipfs_daemon.stop()
+
+
+def pytest_configure(config):
+    """Register the profiling marker to avoid warnings."""
+    config.addinivalue_line("markers", "profiling: filter profiler tests")
```
Comment (DavidMinarsch) [inline tests/conftest.py:1510]: Do that in `pytest.ini` please

File: tests/conftest.py (hunk 3 above)
Comment (DavidMinarsch) [reply tests/conftest.py:1518]: The other approach is to just tag the relevant tests with a marker and then explicitly select the marker. Then you limit it to the marked tests (that's how it's done for integration tests for instance).

Comment (dvilelaf) [reply tests/conftest.py]: Ok, I'll go that route. That was what I've been trying to do but every post I saw talked about adding pytest_collection_modifyitems. I've just talked to Viraj and I'll create a separate job in the workflow.

File: .github/workflows/workflow.yml [hunk 6]
```diff
@@ -304,7 +304,39 @@ jobs:
           unzip protoc-3.19.4-linux-x86_64.zip -d protoc
```
Comment (dvilelaf) [inline .github/workflows/workflow.yml:304]: Note: p2plib tests failing. After checking with Michiel, he confirmed those are flaky ATM.

**Review by DavidMinarsch — APPROVED** [review]

**Comment by Karrenbelt** [comment]
LGTM

**Pull request merged**